distributed/spill.py

```diff
@@ -49,28 +58,53 @@ def disk(self) -> Mapping[Hashable, Any]:
         """
         return self.slow
 
-    @staticmethod
-    def _weight(key: Hashable, value: Any) -> int:
-        return safe_sizeof(value)
-
-    def _on_evict(self, key: Hashable, value: Any) -> None:
-        b = safe_sizeof(value)
-        self.spilled_by_key[key] = b
-        self.spilled_total += b
-
-    def _on_retrieve(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key)
-
-    def __setitem__(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
-        super().__setitem__(key, value)
-        if key in self.slow:
-            # value is individually larger than target so it went directly to slow.
-            # _on_evict was not called.
-            b = safe_sizeof(value)
-            self.spilled_by_key[key] = b
-            self.spilled_total += b
-
-    def __delitem__(self, key: Hashable) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
+    @property
+    def spilled_total(self) -> int:
+        return self.slow.total_weight
+
+
+def _in_memory_weight(key: Hashable, value: Any) -> int:
+    return safe_sizeof(value)
+
+
+class MaxSpillExceeded(Exception):
+    pass
+
+
+class Slow(zict.Func):
+    weight_by_key: dict[Hashable, int]
+    total_weight: int
+
+    def __init__(self, spill_directory: str, max_weight: int | None = None):
+        super().__init__(
+            partial(serialize_bytelist, on_error="raise"),
+            deserialize_bytes,
+            zict.File(spill_directory),
+        )
+        self.max_weight = max_weight
+        self.weight_by_key = {}
+        self.total_weight = 0
+
+    def __setitem__(self, key, value):
+        pickled = self.dump(value)
+        pickled_size = sum(
+            map(safe_sizeof, pickled)
+        )  # self.dump(value)  #this returns a list of len 3 then the size pof pickle is not the value we want
```

Comment:
You don't want to store the in-memory overhead of the bytes objects
```suggestion
        pickled_size = sum(len(frame) for frame in pickled)
```